### PR TITLE
Remove enableRegex

### PR DESCRIPTION
**Pull request by mjfwebb**
# Description

This PR removes `config.enableRegex` as it has been replaced by `MATCHES` in `features.filters`.

## Complexity

Low

### CODE REVIEW

**Comment by neo4j-team-graphql** [comment]
# Performance Report

No Performance Changes


<details>
    <summary>Show Full Table</summary>
    
| name | dbHits | old dbHits | time (ms) | old time (ms) | maxRows |
| ---- | ------ | ---------- | --------- | ------------- | ------- |
|  aggregations.TopLevelAggregate | 3403 | 3403 | 43 | 67 | 1134 |
|  aggregations.NestedAggregation | 14551 | 14551 | 84 | 114 | 2174 |
|  aggregations.AggregationWithWhere | 11979 | 11979 | 57 | 75 | 2174 |
|  aggregations.AggregationWhereWithinNestedRelationships | 22116987 | 22116987 | 3734 | 3724 | 2008534 |
|  aggregations.AggregationWhereWithinNestedConnections | 22116987 | 22116987 | 3557 | 3748 | 2008534 |
|  aggregations.NestedCountFromMovieToActors | 9734 | 9734 | 45 | 69 | 2174 |
|  aggregations.NestedCountFromActorsToMovie | 9937 | 9937 | 59 | 67 | 2174 |
|  aggregations.DeeplyNestedCount | 13070331 | 13070331 | 5625 | 5636 | 2008534 |
|  batch-create.BatchCreate | 4200 | 4200 | 162 | 361 | 600 |
|  connect.createAndConnect | 14424 | 14424 | 226 | 310 | 3003 |
|  connections.Connection | 14082 | 14082 | 80 | 74 | 2174 |
|  connections.NestedConnection | 41462 | 41462 | 203 | 168 | 4516 |
|  connections.ConnectionWithSort | 2282 | 2282 | 82 | 104 | 1040 |
|  connections.ConnectionWithSortAndCypher | 15323 | 15323 | 121 | 157 | 2174 |
|  create.SimpleMutation | 7 | 7 | 67 | 75 | 1 |
|  cypher-directive.TopLevelMutationDirective | 1135 | 1135 | 32 | 51 | 1134 |
|  delete.SimpleDelete | 19401 | 19401 | 328 | 324 | 1040 |
|  delete.NestedDeleteInUpdate | 18196 | 18196 | 305 | 392 | 1179 |
|  2925.SingleRelationshipFilter | 6468 | 6468 | 73 | 51 | 1040 |
|  2925.NestedSingleRelationshipFilter | 22900 | 22900 | 91 | 138 | 2174 |
|  2925.SingleRelationshipRequiredFilter | 3450 | 3450 | 30 | 35 | 1134 |
|  2925.NestedSingleRelationshipRequiredFilter | 9383 | 9383 | 58 | 71 | 1040 |
|  query.SimpleQuery | 15160 | 15160 | 60 | 68 | 2174 |
|  query.QueryWhere | 9715 | 9702 | 40 | 46 | 2167 |
|  query.SimpleQueryWithNestedWhere | 9893 | 9880 | 58 | 78 | 2167 |
|  query.Nested | 10092037 | 10092037 | 11989 | 12447 | 2008534 |
|  query.NestedWithFilter | 40401 | 40401 | 134 | 171 | 4511 |
|  query.OrFilterOnRelationships | 42261 | 42907 | 206 | 263 | 1637 |
|  query.OrFilterOnRelationshipsAndNested | 36360 | 36410 | 253 | 498 | 1637 |
|  query.QueryWithNestedIn | 13473 | 13224 | 67 | 70 | 1241 |
|  query.NestedConnectionWhere | 9834 | 9834 | 70 | 71 | 2174 |
|  query.DeeplyNestedConnectionWhere | 10038 | 9973 | 98 | 109 | 2174 |
|  query.DeeplyNestedWithRelationshipFilters | 19284 | 19230 | 160 | 208 | 1642 |
|  query.NestedWithRelationshipSingleFilters | 3024400 | 3024400 | 883 | 811 | 1003942 |
|  query.Fulltext | 96 | 96 | 29 | 59 | 16 |
|  query.FulltextWithNestedQuery | 587 | 587 | 49 | 56 | 84 |
|  sorting.SortMultipleTypes | 2513 | 2513 | 78 | 85 | 1040 |
|  sorting.SortMultipleTypesWithCypherWithCypher | 14585 | 14546 | 108 | 134 | 2174 |
|  sorting.SortOnNestedFields | 14082 | 14082 | 58 | 63 | 2174 |
|  sorting.SortDeeplyNestedFields | 42196 | 42196 | 100 | 126 | 4516 |
|  sorting.SortWithTopLevelCypher | 15160 | 15160 | 65 | 70 | 2174 |
|  unions.SimpleUnionQuery | 321 | 321 | 73 | 79 | 35 |
|  unions.SimpleUnionQueryWithMissingFields | 293 | 293 | 87 | 80 | 35 |
|  unions.NestedUnion | 410637 | 410637 | 336 | 353 | 33033 |
|  unions.NestedUnionWithMissingFields | 384611 | 384611 | 330 | 377 | 33033 |
|  update.NestedUpdate | 2119 | 2119 | 75 | 80 | 1040 |

</details>
        
Old Schema Generation: 42.650s
Schema Generation: 42.769s
Old Subgraph Schema Generation: 59.083s
Subgraph Schema Generation: 59.612s

File: packages/graphql/src/schema/get-where-fields.ts [hunk 1]
```diff
@@ -126,30 +124,25 @@ function getWhereFields({
...
-                    if (matchesSetting === true) {
-                        res[`${f.fieldName}_MATCHES`] = { type: "String", directives: deprecatedDirectives };
-                    }
-                } else if (enableRegex) {
-                    // TODO: This is deprecated. To be removed in 4.0.0.
-                    res[`${f.fieldName}_MATCHES`] = { type: "String", directives: deprecatedDirectives };
-                }
-
-                const stringWhereOperators = ["_CONTAINS", "_STARTS_WITH", "_ENDS_WITH"];
+                const stringWhereOperators: { comparator: string; typeName: string }[] = [
```
Comment (angrykoala) [inline packages/graphql/src/schema/get-where-fields.ts:127]: **nitpick, not important**: For these complex object I find array types to be a bit more readable as
 ```ts
Array<{ comparator: string; typeName: string }>
```

File: packages/graphql/src/schema/get-where-fields.ts [hunk 2]
```diff
@@ -126,30 +124,25 @@ function getWhereFields({
...
+                    { comparator: "_ENDS_WITH", typeName: f.typeMeta.name },
+                ];
 
                 const stringWhereOperatorsNegate = ["_NOT_CONTAINS", "_NOT_STARTS_WITH", "_NOT_ENDS_WITH"];
 
-                Object.entries(features?.filters?.String || {}).forEach(([filter, enabled]) => {
-                    if (filter === "MATCHES") {
-                        return;
-                    }
+                Object.entries(features?.filters?.[f.typeMeta.name] || {}).forEach(([filter, enabled]) => {
```
Comment (angrykoala) [inline packages/graphql/src/schema/get-where-fields.ts:135]: **nitpick** Consider creating a varaible for `f.typeMeta.name` as it is used in several places and the extra context of a explicit name could help with readability here

File: .changeset/wicked-ties-burn.md [hunk 3]
```diff
@@ -0,0 +1,5 @@
+---
+"@neo4j/graphql": major
+---
+
+enableRegex has been removed and replaced with MATCHES filters in the features configuration object. See the migration guide for more information: https://neo4j.com/docs/graphql-manual/current/guides/v4-migration
```
Comment (darrellwarde) [inline .changeset/wicked-ties-burn.md:5]:
```suggestion
`enableRegex` has been removed and replaced with `MATCHES` filters in the features configuration object. See the migration guide for more information: https://neo4j.com/docs/graphql-manual/current/guides/v4-migration
```

File: docs/modules/ROOT/pages/guides/v4-migration/index.adoc [hunk 4]
```diff
@@ -19,6 +19,29 @@ npm update @neo4j/graphql
 If you were passing any arguments from https://the-guild.dev/graphql/tools/docs/api/interfaces/schema_src.iexecutableschemadefinition[`IExecutableSchemaDefinition`] into the library
 other than `typeDefs` and `resolvers`, these are no longer supported.
 
+=== config.enableRegex replaced by `MATCHES` in features.filters
+
+config.enableRegex has been replaced by `MATCHES` in features.filters. With this change comes more granularity in the feature configuration. You can now enable the `MATCHES` filter on `String` and `ID` fields separately.
```
Comment (darrellwarde) [inline docs/modules/ROOT/pages/guides/v4-migration/index.adoc:24]:
```suggestion
`config.enableRegex` has been replaced by `MATCHES` in features.filters. With this change comes more granularity in the feature configuration. You can now enable the `MATCHES` filter on `String` and `ID` fields separately.
```

File: packages/graphql/src/schema/get-where-fields.ts [hunk 5]
```diff
@@ -126,30 +124,25 @@ function getWhereFields({
...
             if (["String", "ID"].includes(f.typeMeta.name)) {
-                const matchesSetting: boolean | undefined = features?.filters?.[f.typeMeta.name]?.MATCHES;
-                if (matchesSetting !== undefined) {
-                    if (matchesSetting === true) {
-                        res[`${f.fieldName}_MATCHES`] = { type: "String", directives: deprecatedDirectives };
-                    }
-                } else if (enableRegex) {
-                    // TODO: This is deprecated. To be removed in 4.0.0.
-                    res[`${f.fieldName}_MATCHES`] = { type: "String", directives: deprecatedDirectives };
-                }
```
Comment (darrellwarde) [inline packages/graphql/src/schema/get-where-fields.ts:137]: Much tidier! 👏

**Review by darrellwarde — APPROVED** [review]

File: packages/graphql/src/schema/get-where-fields.ts (hunk 2 above)
Comment (angrykoala) [inline packages/graphql/src/schema/get-where-fields.ts:135]: Also, I think we could really make use of a `features` helper here, as this seems like it is going to be the way forwards for features flag and we should make our lifes a bit easier on this I think.

I'm thinking something a bit like:
```ts
features(f.typeMeta.name).hasFilter("MATCHES");
```

This way we have a more unified way of accesing features that don't require concatenation of `.?` and default objects and we could encapsulate the logic of traversing the filters, avoiding this `forEach` on our schema generation logic.

Thoughts?

**Review by angrykoala — COMMENTED** [review]
Looks good, just one important comment that we should either tackle or discuss before merging

**Pull request merged**